server/keyspace/keyspace.go

```diff
@@ -0,0 +1,306 @@
+	// First get keyspace ID from the name given.
+	loaded, spaceID, err := manager.store.LoadKeyspaceIDByName(name)
+	if err != nil {
+		return nil, err
+	}
+	if !loaded {
+		return nil, ErrKeyspaceNotFound
+	}
+	return manager.loadKeyspaceByID(spaceID)
+}
+
+func (manager *Manager) loadKeyspaceByID(spaceID uint32) (*keyspacepb.KeyspaceMeta, error) {
+	// Load the keyspace with target ID.
+	keyspace := &keyspacepb.KeyspaceMeta{}
+	loaded, err := manager.store.LoadKeyspace(spaceID, keyspace)
+	if err != nil {
+		return nil, err
+	}
+	if !loaded {
+		return nil, ErrKeyspaceNotFound
+	}
+	return keyspace, nil
+}
+
+// Mutation represents a single operation to be applied on keyspace config.
+type Mutation struct {
+	Op    OpType
+	Key   string
+	Value string
+}
+
+// OpType defines the type of keyspace config operation.
+type OpType int
+
+const (
+	// OpPut denotes a put operation onto the given config.
+	// If target key exists, it will put a new value,
+	// otherwise, it creates a new config entry.
+	OpPut OpType = iota + 1 // Operation type starts at 1.
+	// OpDel denotes a deletion operation onto the given config.
+	// Note: OpDel is idempotent, deleting a non-existing key
+	// will not result in error.
+	OpDel
+)
+
+// UpdateKeyspaceConfig changes target keyspace's config in the order specified in mutations.
+// It returns error if saving failed, operation not allowed, or if keyspace not exists.
+func (manager *Manager) UpdateKeyspaceConfig(name string, mutations []*Mutation) (*keyspacepb.KeyspaceMeta, error) {
+	// First get KeyspaceID from Name.
+	loaded, spaceID, err := manager.store.LoadKeyspaceIDByName(name)
+	if err != nil {
+		return nil, err
+	}
+	if !loaded {
+		return nil, ErrKeyspaceNotFound
+	}
+	manager.metaLock.Lock(spaceID)
+	defer manager.metaLock.Unlock(spaceID)
+	// Load keyspace by id.
+	keyspace, err := manager.loadKeyspaceByID(spaceID)
+	if err != nil {
+		return nil, err
+	}
+	// Changing ARCHIVED keyspace's config is not allowed.
+	if keyspace.State == keyspacepb.KeyspaceState_ARCHIVED {
+		return nil, errKeyspaceArchived
+	}
+	if keyspace.Config == nil {
+		keyspace.Config = map[string]string{}
+	}
+	// Update keyspace config according to mutations.
+	for _, mutation := range mutations {
+		switch mutation.Op {
+		case OpPut:
+			keyspace.Config[mutation.Key] = mutation.Value
+		case OpDel:
+			delete(keyspace.Config, mutation.Key)
+		default:
+			return nil, errIllegalOperation
+		}
+	}
+	// Save the updated keyspace.
+	if err = manager.store.SaveKeyspace(keyspace); err != nil {
+		return nil, err
+	}
+	return keyspace, nil
+}
+
+// UpdateKeyspaceState updates target keyspace to the given state if it's not already in that state.
+// It returns error if saving failed, operation not allowed, or if keyspace not exists.
+func (manager *Manager) UpdateKeyspaceState(name string, newState keyspacepb.KeyspaceState, now time.Time) (*keyspacepb.KeyspaceMeta, error) {
```

Comment:
Why do we need to pass now as a parameter?

Reply:
Mostly for two reasons:
1. It makes it easier to generate test data for future time related functionalities (For example, archiving keyspace that has been disabled for more than x days)
2. It's possible that now is parsed from tso rather than from `time.Now` directly. (See [`UpdateServiceGCSafePoint`](https://github.com/tikv/pd/blob/e089172a12c221655044b92210acc0477cf9ccb2/server/grpc_service.go#L1434))

But I'm not sure how necessary it is to used TSO to record keyspace info. So I'm open to remove it from parameter and generate it at the manager.